Merge from upstream v13

Alchez opened the pull request:
Changelogs:

- https://github.com/frappe/frappe/pull/15557
- https://github.com/frappe/frappe/pull/15689
- https://github.com/frappe/frappe/pull/15924

navdeepghai committed:
feat(fixes): for frappeclient post_api(POST) request being redirected to GET request on calling server

(cherry picked from commit d73428df919dc98e7ee13af92af5fd8a66da66d4)

gavindsouza committed:
refactor: frappe.db.set_value

* Simplified logic
* Perf enhancements (removed unnecessary conditional computations)
* Use query builder and ORM to build queries instead of juggling

(cherry picked from commit bb29d417ed8eb289564282c8be70ab8076df2a9b)

gavindsouza committed:
chore: Bump psycopg2-binary Python client

(cherry picked from commit 39d63641f3d9928b2648ec83e793badf5c92f544)

gavindsouza committed:
fix: Expand iterable to get rid of extra brackets

Although valid SQL, MariaDB didn't like double brackets on VALUES. It
raised ERROR 1241: Operand should contain 1 column(s)

(cherry picked from commit 4b0b07c07933893cb605752a4f467c65571221c5)

gavindsouza committed:
fix: Run query even if no documents found for_update

If no docnames are found, then use NULL as fallback. I tested this on PostgreSQL & MariaDB and it seems to work as expected

    In [1]: from frappe.query_builder import Field

    In [2]: frappe.db.set_value('ToDo', Field("creation") > Field("modified"), 'description', 'change 2', for_update=True, debug=True)
    SELECT "name" FROM "tabToDo" WHERE "creation">"modified"
    Execution time: 0.0 sec
    UPDATE "tabToDo" SET "description"='change 2',"modified"='2021-10-22 09:38:40.110481',"modified_by"='Administrator' WHERE "name" IN (NULL)
    Execution time: 0.0 sec

(cherry picked from commit d0f71df84f1315c902c155bfa9d26b8495e0ab5f)

gavindsouza committed:
fix: Cast values as str for all single doctypes

Re-arranged block for simplicity. Type casting doesn't change anything
it seems. On get_value, values are casted properly.

(cherry picked from commit a7187d44d64e048ce404e04f4568959e12392dbf)

saxenabhishek committed:
fix: misc setup_wizard region slide

placeholder on all fields
autocomplete on country input
default selected values on inputs

(cherry picked from commit 91bd5ed2e2a8413574ac26e0620157d59022cc73)

Commit:
fix: remove return statement for empty link field

surajshetty3416 committed:
refactor: Use placeholder instead of additional option in select

- Additional options were still selectable which could have got through
- Minor formatting changes

(cherry picked from commit 1dd870ce73457b1505e74c706be14ea8a7a33867)

surajshetty3416 committed:
revert: Empty value checks

(cherry picked from commit b530beb503c0f70422daafb98817741c24ffd21e)

shariquerik committed:
fix: do not validate if options is undefined

pruthvi145 committed:
fix: clear cache when filters passed as dn param in `set_value`

(cherry picked from commit 289d7e7afa8c922d9069579d3a898ba98b529a9b)

surajshetty3416 committed:
refactor: Re-arraange fields in user permission to avoid confusion

(cherry picked from commit b2d16e7db8af27319bcbeed37e594483dabafe9c)
(cherry picked from commit a58fe1a9e7434636ff089b26e6c57f094fa2d8f5)

surajshetty3416 committed:
refactor: Re-arrange user form

- To fix order of fields in quick entry form

(cherry picked from commit 0868ad02f84ce5a3593fa9d89fbcf597887ac6fc)

# Conflicts:
#	frappe/core/doctype/user/user.json

surajshetty3416 committed:
Merge pull request #15500 from frappe/mergify/bp/version-13-pre-release/pr-15499

surajshetty3416 committed:
fix: Start week from Sunday instead of Monday

- Like we do in client-side

(cherry picked from commit b5e2f78a883d13e5acd477130aab7dc59e1fc0c2)

surajshetty3416 committed:
test: Add cases to check date utilities

(cherry picked from commit 85360ef4c0c4204cf79c6248e122924993310f19)

# Conflicts:
#	frappe/tests/test_utils.py

surajshetty3416 committed:
fix: Do not allow rebuilding of tree for doctypes without `lft` & `rgt`

(cherry picked from commit 9258941151616c7099c09564975a70c9cd3bc642)

surajshetty3416 committed:
fix: Hide "Rebuild Tree" option if treeview is loaded for non-nestedset doctype

(cherry picked from commit d1ca3fa78c69aefb3da20470cc0537fc83e0b179)

surajshetty3416 committed:
fix: Use copy of an array while adding new content for proper change trigger

(cherry picked from commit 3e009e9bdb95ab8b4f684e4bc8a30829fb8ae6b5)

nextchamp-saqib committed:
fix: renaming a document updates the modified timestamp

(cherry picked from commit b0b19d7a09d6233e0b14c28b805fe893c528ff99)

ankush committed:
Merge pull request #15510 from frappe/mergify/bp/version-13-hotfix/pr-15508

fix: renaming a document updates the modified timestamp (backport #15508)

ankush committed:
Merge pull request #15512 from frappe/mergify/bp/version-13-hotfix/pr-15504

fix: Do not allow rebuilding of tree for doctypes without `lft` & `rgt` (backport #15504)

surajshetty3416 committed:
Merge pull request #15513 from frappe/mergify/bp/version-13-hotfix/pr-15507

surajshetty3416 committed:
fix: Resolve conflicts

surajshetty3416 committed:
Merge pull request #15514 from frappe/mergify/bp/version-13-hotfix/pr-15498

mergify[bot] committed:
Merge pull request #15477 from deepeshgarg007/query_report_route_options

fix: Route options not getting set while navigating from one report to another

surajshetty3416 committed:
feat: Add setting to configure the day on which week starts

(cherry picked from commit 42e1c15c18f72e447f1b7b5e4227963abbc90249)

# Conflicts:
#	frappe/core/doctype/system_settings/system_settings.json
#	frappe/public/js/frappe/desk.js

surajshetty3416 committed:
fix: Set start of the week for date control based on configuration

(cherry picked from commit 9c75336ea0e11d31adcf23b129f351d148b2569a)

surajshetty3416 committed:
test: Update date utils test cases

(cherry picked from commit 4a5d9a02e3b74b5c430f45514dc6fc5ccb3b7aa8)

surajshetty3416 committed:
fix: Add patch to set week_starts_on as "Monday"

(cherry picked from commit ca9b2a953d13a3c2b67a37c9aaba2da0d93fb8cc)

# Conflicts:
#	frappe/patches.txt

surajshetty3416 committed:
test: Fix dashboard chart tests

(cherry picked from commit 5821a7bc9176ad8fb11d39bf89212d29066f327d)

surajshetty3416 committed:
style: Fix formatting issues

(cherry picked from commit ea2101d537cab6541f047f9fe1879f8303474a5a)

surajshetty3416 committed:
fix: Revert unnecessary code

(cherry picked from commit 6b77ed68e9b74b85cf5bbee1c05630a8e36d2ebb)

surajshetty3416 committed:
fix: Re-setup moment object whenever `week_starts_on` is changed

(cherry picked from commit 53b2c472a6f4576ef7fbeea478d54b1ec54ba7ae)

# Conflicts:
#	frappe/core/doctype/system_settings/system_settings.js

mergify[bot] committed:
feat(minor): copy link to clipboard (backport #15523) (#15524)

Co-authored-by: Ankush Menat <ankush@frappe.io>

mergify[bot] committed:
Merge pull request #15530 from frappe/mergify/bp/version-13-hotfix/pr-15492

fix: clear cache for all docs matched with the filters passed as `dn` param in `set_value` (backport #15492)

surajshetty3416 committed:
refactor: Rename `week_starts_on` to `first_day_of_the_week`

For consistency

(cherry picked from commit 5d5ad78789738e6c9ddd3a368f44badc9022ce5a)

mergify[bot] committed:
fix(ldap): Auto-negotiate the highest TLS protocol version supported by client & server (backport #15517) (#15535)

Co-authored-by: Christoph Kappel <christoph.kappel@bytesource.net>

mergify[bot] committed:
fix: set `first_response_time` only if communication is sent (backport #15533) (#15536)

Co-authored-by: Saqib Ansari <nextchamp.saqib@gmail.com>

nextchamp-saqib committed:
feat: `frappe.enqueue` and `frappe.call` for server scripts (#15528)

Co-authored-by: Sagar Vora <sagar@resilient.tech>

gavindsouza committed:
fix: Use params instead of data for client login

(cherry picked from commit 984420363a9aa8d60f95b197e0750e1766347ddf)

# Conflicts:
#	frappe/frappeclient.py

gavindsouza committed:
Merge pull request #15540 from frappe/mergify/bp/version-13-hotfix/pr-14323

chore: Bump psycopg2-binary Python client (backport #14323)

barredterra committed:
fix: awesomeplete line wrapping

(cherry picked from commit bf28199a6dd14c63ac7feb0ec1ace67eeff57408)

surajshetty3416 committed:
test: UI test cases for "First Day of the Week"

(cherry picked from commit 54fbd0b5e6239398e0833557e718ff5bf96d578d)

surajshetty3416 committed:
style: Fix sider issues

(cherry picked from commit ed2e37eaf7c61e0687d9d817505b10b2654c8f4a)

surajshetty3416 committed:
test: Fix flaky grid flaky test

Simplified test case as bit

(cherry picked from commit 30591720f868e924973c2fd458448752d8a5295c)

surajshetty3416 committed:
test: Remove unnecessary code

(cherry picked from commit 454a08aa183d81cbaed1c3ff789a5943c92ddd82)

mergify[bot] committed:
fix(Migrate): add unique and index for new columns (backport #15484) (#15545)

Co-authored-by: Saif Ur Rehman <saifi0102@gmail.com>

gavindsouza committed:
fix: Use format instead of f-strings, resolve conflicts

gavindsouza committed:
Merge pull request #15546 from frappe/mergify/bp/version-13-hotfix/pr-15542

fix: FrappeClient post_api request being redirected to GET request (backport #15542)

nextchamp-saqib committed:
fix: catch exceptions while parsing json on request failure

(cherry picked from commit 55178101e85ec80b367cd88f84e251c4278c61ca)

surajshetty3416 committed:
test: Fix get_field command for select fields

(cherry picked from commit 0a8b5fe7857c875c6b08b728887d93ba96dd2cd4)

ahmadpak committed:
fix: email_string parsing when pulling email

(cherry picked from commit 6afa5c8da3c7d0e74425f572e39f042e76ae6d19)

# Conflicts:
#	frappe/core/doctype/communication/communication.py

surajshetty3416 committed:
fix: Resolve conflicts

ahmadpak committed:
refactor: communication.py

(cherry picked from commit 9ebfc1ced9cfb7643560858553e59243eb128ac4)

surajshetty3416 committed:
Merge pull request #15555 from frappe/mergify/bp/version-13-hotfix/pr-15502

rohitwaghchaure committed:
chore: change log for v13.18.0

rohitwaghchaure committed:
Merge pull request #15557 from rohitwaghchaure/change-log-for-v13-18

chore: change log for v13.18.0

rohitwaghchaure committed:
bumped to version 13.18.0

rohitwaghchaure committed:
Merge branch 'version-13-pre-release' into version-13

surajshetty3416 committed:
Merge branch 'version-13-hotfix' into mergify/bp/version-13-hotfix/pr-15401

surajshetty3416 committed:
fix: translations in setup wizard (backport #15401) (#15473)

Co-authored-by: Raffael Meyer <14891507+barredterra@users.noreply.github.com>
Co-authored-by: Himanshu <himanshuwarekar@yahoo.com>
Co-authored-by: Suraj Shetty <13928957+surajshetty3416@users.noreply.github.com>

surajshetty3416 committed:
Merge pull request #15558 from frappe/mergify/bp/version-13-hotfix/pr-15543

gavindsouza committed:
fix: Clear cache on set_value

(cherry picked from commit 16377ad4f3d93a559f062e98101abed5085a35a9)

Commit:
Merge remote-tracking branch 'upstream/version-13' into merge-v13

gavindsouza committed:
fix(typing): Add type hints for frappe.cache

(cherry picked from commit 46ea4022d0050663fad9c737a15c7923c40b2eaf)

ghost commented:
### Code Refactoring

* Re-arraange fields in user permission to avoid confusion ([c66dcb5](https://github.com/ParsimonyGit/frappe/commit/c66dcb5eadfbdd991884f4d28f2661dd0e165869))
* Re-arrange user form ([cdbfd80](https://github.com/ParsimonyGit/frappe/commit/cdbfd808b9c96e05f9b1e8140ae6049503a6375f))


### Bug Fixes

* renaming a document updates the modified timestamp ([e538038](https://github.com/ParsimonyGit/frappe/commit/e538038a3546df8700f641bb1dc472daad7dd3b6))
* Do not allow rebuilding of tree for doctypes without `lft` & `rgt` ([0c11f1a](https://github.com/ParsimonyGit/frappe/commit/0c11f1ac390bacc5782c0e9854d67c5e369ec2b7))
* Hide "Rebuild Tree" option if treeview is loaded for non-nestedset doctype ([59b3f19](https://github.com/ParsimonyGit/frappe/commit/59b3f19bf4923a0f8bd5db1591f3f8a297ebda5f))
* Use copy of an array while adding new content for proper change trigger ([c7ef015](https://github.com/ParsimonyGit/frappe/commit/c7ef01580250472ab66d515c17cfc0fb78977308))
* Resolve conflicts ([74f5064](https://github.com/ParsimonyGit/frappe/commit/74f50642a966c57a223790400f74e93ba402c568))
* clear cache when filters passed as dn param in `set_value` ([2ffb2c9](https://github.com/ParsimonyGit/frappe/commit/2ffb2c954ad1515ca9417330c7db88727d69eeae))
*  **ldap:** Auto-negotiate the highest TLS protocol version supported by client & server (backport #15517) (#15535) ([1b69b39](https://github.com/ParsimonyGit/frappe/commit/1b69b3978f49efd308614ce1311e98529b471429))
* set `first_response_time` only if communication is sent (backport #15533) (#15536) ([f561714](https://github.com/ParsimonyGit/frappe/commit/f561714b10cb52a0b7d2a95f253c6fbbfa072ba0))
*  **Migrate:** add unique and index for new columns (backport #15484) (#15545) ([0679b99](https://github.com/ParsimonyGit/frappe/commit/0679b99146b0450866ca6f674533ad2a71477aed))
* Use params instead of data for client login ([8d73b16](https://github.com/ParsimonyGit/frappe/commit/8d73b1615f9610033ee7a3f979a86793f40f4274))
* Use format instead of f-strings, resolve conflicts ([2787bb8](https://github.com/ParsimonyGit/frappe/commit/2787bb8597f72cdc6242dd4bcdb62289db60c61c))
* Start week from Sunday instead of Monday ([3503015](https://github.com/ParsimonyGit/frappe/commit/350301551fa96e7dc6a80b884673cc95065d6905))
* Set start of the week for date control based on configuration ([5e64c15](https://github.com/ParsimonyGit/frappe/commit/5e64c151c3a6b958c01bf85ffcfdd13c82d83859))


### Features

*  **minor:** copy link to clipboard (backport #15523) (#15524) ([356f7a8](https://github.com/ParsimonyGit/frappe/commit/356f7a8509582a5067f60d44b01df5956df3b480))
* `frappe.enqueue` and `frappe.call` for server scripts (#15528) ([732d776](https://github.com/ParsimonyGit/frappe/commit/732d7760aa49f7faf4d4f8526dafa81bc43e9ea0))
*  **fixes:** for frappeclient post_api(POST) request being redirected to GET request on calling server ([0942bb6](https://github.com/ParsimonyGit/frappe/commit/0942bb675ab3a374b434d4ef68311dc6b8b2e71c))
* Add setting to configure the day on which week starts ([c8dd803](https://github.com/ParsimonyGit/frappe/commit/c8dd803d2d2c3d98dd17d58ce31270ab914443c4))


### Chore

* Bump psycopg2-binary Python client ([a0b0c10](https://github.com/ParsimonyGit/frappe/commit/a0b0c10321edc8a2b4aa8b830a0b804a8201dcba))


### Tests

* Add cases to check date utilities ([17268c3](https://github.com/ParsimonyGit/frappe/commit/17268c3f4df02f939e97e541e5139cdb68330e77))


### Contributors

<a href="https://github.com/surajshetty3416">surajshetty3416</a>, <a href="https://github.com/nextchamp-saqib">nextchamp-saqib</a>, <a href="https://github.com/ankush">ankush</a>, <a href="https://github.com/mergify">mergify[bot]</a>, <a href="https://github.com/pruthvi145">pruthvi145</a>, <a href="https://github.com/gavindsouza">gavindsouza</a>, <a href="https://github.com/navdeepghai1">navdeepghai1</a>

<details>
<summary>Commit-Lint commands</summary>
<br />

You can trigger Commit-Lint actions by commenting on this PR:
- `@Commit-Lint merge patch` will merge dependabot PR on "patch" versions (X.X.Y - Y change)
- `@Commit-Lint merge minor` will merge dependabot PR on "minor" versions (X.Y.Y - Y change)
- `@Commit-Lint merge major` will merge dependabot PR on "major" versions (Y.Y.Y - Y change)
- `@Commit-Lint merge disable` will desactivate merge dependabot PR
- `@Commit-Lint review` will approve dependabot PR
- `@Commit-Lint stop review` will stop approve dependabot PR



</details>

nextchamp-saqib committed:
chore: check custom exception handlers only for json response

(cherry picked from commit 307c7ced69572ec56051f3a88e1e91422d2e94fb)

nextchamp-saqib committed:
fix: parsing exception from traceback

(cherry picked from commit 1bde3b48bdd7a0d81f6d183b19f58ac8ecb8b72c)

gavindsouza committed:
fix(qb): Patch ParameterizedValueWrapper as wrapper_cls

(cherry picked from commit 19794ba2730e6645558a8741d3411cdbdc21e9bd)

gavindsouza committed:
style: Sorted imports & whitespace consistency

(cherry picked from commit 2c2d200728483102d6ed131800cba8ba36ab71ea)

pruthvi145 committed:
fix: delete translation only if label is empty

(cherry picked from commit 97edb8f267794253c20216951e75d49102cf7814)

gavindsouza committed:
style: remove commented code, whitespaces test_file, added typing hint

(cherry picked from commit b4c7c2bc55034623e811a0ee84fa7cf59f26acf0)

# Conflicts:
#	frappe/core/doctype/file/test_file.py

gavindsouza committed:
fix(qb:ValueWrapper): Use get_value_sql only for str values

(cherry picked from commit 2fcecef10a34bada8a8e7ecacec77ca3e17ad3d6)

gavindsouza committed:
chore(typing): Add type hints in qb builder classes

(cherry picked from commit 5a289f41e74a825108e664ef2737059f2080c6fb)

gavindsouza committed:
fix: Don't cast to str if None or Falsy

* refactor: use get_single_value instead of set_value(blah, None, blah1)
* User.reload in test_home_page to combat what happens locally

(cherry picked from commit c2adb16233b65753051f9844fd2a5524989e2308)

ankush committed:
revert: "fix: translations in setup wizard (backport #15401)" (#15572)

pruthvi145 committed:
fix: limit in `Document.get`

(cherry picked from commit d1852242a0acb37c799d00c2b9fc8ca6d71b3587)

# Conflicts:
#	frappe/model/base_document.py

saxenabhishek committed:
feat: build constant value cols

* Ported https://github.com/frappe/frappe/pull/15212
* styled test_query_builder

(cherry picked from commit f310e2c12035120ab0e9368aae1f9339fa14ac2a)

gavindsouza committed:
fix(single:set_value): Cast using sbool instead of str for all

(cherry picked from commit 1c367ec1c305a2bc44bdb48e855a6f2ce7a75618)

gavindsouza committed:
test: Add tests for frappe.db.set_value

(cherry picked from commit 2f4e06452f94785ceab6038d3b038fedf7cf21f9)

gavindsouza committed:
feat: frappe.db.set_single_value

Alias to frappe.db.set_value without having to pass the doctype name
twice or teh second one as None

Other changes:
* set/get db APIs for single DocTypes
* Make use of redundant cache key in frappe.db.get_single_value

(cherry picked from commit 80464c2224716adff01e3485bf1a37130fc5732d)

saxenabhishek committed:
test: qb parameterization

where conditions
set update conditions
where with a function

(cherry picked from commit 1e0979bfe79c60b8443a899e2165c9095c6d57e2)

saxenabhishek committed:
refactor: pythonic NamedParameterWrapper

(cherry picked from commit 7a36da571bdf80f27618bdb3d72b5aceb6098dee)

saxenabhishek committed:
refactor: remove reduandant func call

(cherry picked from commit 0ea0237206f2b8091600bc1f69628b8911eca639)

saxenabhishek committed:
docs: Working of NamedParameters in qb

(cherry picked from commit 02088b5e96b641ba83c6f11b1d28be2765f4f52a)

pruthvi145 committed:
test: add test for limit

(cherry picked from commit a87899c2155a0c2576d551a31aa5f6535be6930b)

gavindsouza committed:
test: Postgres compatible db tests

(cherry picked from commit 2af4a20af800df9ea8afbb8ed90fb84965232aef)

gavindsouza committed:
fix: sbool before cint in cast for Check types

Noticed an issue when get_single_value wasn't returning the correct
values; by converting true to 0.

Tested it out. Here's the examples:

In [2]: sbool("2")
Out[2]: '2'

In [3]: cint(sbool("2"))
Out[3]: 2

In [4]: cint(sbool("-1"))
Out[4]: -1

In [5]: cint(sbool("0"))
Out[5]: 0

In [6]: cint(sbool("1000"))
Out[6]: 1000

In [7]: cint(sbool("10_000"))
Out[7]: 10000

In [8]: cint(sbool("true"))
Out[8]: 1

(cherry picked from commit 6b66f9320e05c8686d1e07258cd147d8c76e93e6)

sumaiya2908 committed:
fix: email fetch error for imaplib (#15581)

* fix: remove maxline for imaplib

* refactor: remove redundant line

Co-authored-by: Summayya <frappe@Summayyas-MacBook-Air.local>
(cherry picked from commit 645fde2e675daf0fd7e13ad44a051a5453c2577c)

# Conflicts:
#	frappe/email/receive.py

mergify[bot] committed:
Merge pull request #15577 from frappe/mergify/bp/version-13-hotfix/pr-15564

fix: communication method "get_contacts" (backport #15564)

mergify[bot] committed:
Merge pull request #15583 from frappe/mergify/bp/version-13-hotfix/pr-15581

fix: email fetch error for imaplib (backport #15581)

mergify[bot] committed:
Merge pull request #15585 from frappe/mergify/bp/version-13-hotfix/pr-15574

fix: limit in `Document.get` (backport #15574)

saxenabhishek committed:
fix: replace the field method with a column

(cherry picked from commit fa1131dc140e64c51cd5cdd6dc92ba1eeadcd0ac)

gavindsouza committed:
fix: Set cache=True to maintain old behaviour consistency

(cherry picked from commit ea2d6c00309c723950d1872a24f5f5f4b35f306a)

saxenabhishek committed:
fix: ignore copy of getattr methods

(cherry picked from commit 854e0ac2c9f1c65b5fae4a85efb7bc355beac3a5)

mergify[bot] committed:
fix: enable languages inserted during setup (#15588) (#15590)

(cherry picked from commit eeff2a22f88b8f836a4cf0bc5935730318f58108)

Co-authored-by: Ankush Menat <ankush@frappe.io>

mergify[bot] committed:
Merge pull request #15595 from frappe/mergify/bp/version-13-hotfix/pr-15549

fix: catch exceptions while parsing json on request failure (backport #15549)

ruchamahabal committed:
fix(minor): Is Custom field placement in Role form

(cherry picked from commit b878b71e3797e2e4cb66073d8e8eadcd3a725660)

# Conflicts:
#	frappe/core/doctype/role/role.json

surajshetty3416 committed:
fix: Add toggle theme as standard dropdown option

- For new sites toggle theme was not available. It was added via patch for old sites.

(cherry picked from commit 645f23a9a7c28dfa0271888e419ff13b4b4b2986)

surajshetty3416 committed:
Merge pull request #15611 from frappe/mergify/bp/version-13-hotfix/pr-15610

surajshetty3416 committed:
Merge pull request #15573 from frappe/mergify/bp/version-13-hotfix/pr-15277

mergify[bot] committed:
fix: AttributeError when calling frappe.permissions.$function (backport #15615) (#15616)

Co-authored-by: Ankush Menat <ankush@frappe.io>

saxenabhishek committed:
fix: make parameters for strings only

(cherry picked from commit 730917f8628495d4e9a932254ce3f5b6e72769cb)

mergify[bot] committed:
fix: patch for validating options field (#15565) (#15623)

* fix: patch for validating options field

* fix: reorder patch execution

* refactor: reduce queries for patching

* refactor: change patch name and license

Co-authored-by: Ankush Menat <ankush@frappe.io>
(cherry picked from commit c3409c818eba2e0d4e30bf27a8bd8aabadd4c566)

Co-authored-by: Shadrak Gurupnor <30501401+shadrak98@users.noreply.github.com>

gavindsouza committed:
fix!: Don't use strtobool in sbool util

strtobool would convert 't', 'yes' to True too. Which shouldn't be a
problem generally but could be one possibly.

(cherry picked from commit 8691b23a1038fa28fe8693111fd3e9230ef3268a)

gavindsouza committed:
style: Add typing, sorted imports#

(cherry picked from commit 3eff49c661d701c52efa4420790326f80f698582)

shariquerik committed:
build: Upgrade Frappe Gantt from 0.5.0 to 0.6.0 (#15635)

(cherry picked from commit 3f7ede278df62a6a5b63802cdbe4475dd4060683)

# Conflicts:
#	package.json

mergify[bot] committed:
fix: translations in setup wizard (#15401) (#15633)

Co-authored-by: Raffael Meyer <14891507+barredterra@users.noreply.github.com>
(cherry picked from commit 8947a542d1d3152dbf68c6fcc6b22286d01dd52b)

Co-authored-by: Himanshu <himanshuwarekar@yahoo.com>

skjbulcher committed:
feat: Add click-drag scrolling to kanban boards

(cherry picked from commit 115bb8faedae45031b5b08093552fd2d48c483f4)

rohitwaghchaure committed:
chore: Merge branch 'version-13-hotfix' into 'version-13-pre-release'

rohitwaghchaure committed:
Merge pull request #15642 from rohitwaghchaure/merge-hotfix-to-pre-release-for-13-19

chore: Merge branch 'version-13-hotfix' into 'version-13-pre-release'

surajshetty3416 committed:
style: Fix formatting

(cherry picked from commit a6307638498fb181c2dab8207c501a21f4bd34d7)

mergify[bot] committed:
fix: unhide context field (backport #15485) (#15640)

Co-authored-by: hrwx <himanshuwarekar@yahoo.com>

surajshetty3416 committed:
Merge pull request #15643 from frappe/mergify/bp/version-13-hotfix/pr-15641

mergify[bot] committed:
fix: datatable column resize in rtl (backport #15636) (#15644)

Co-authored-by: Faris Ansari <netchamp.faris@gmail.com>
Co-authored-by: Suraj Shetty <surajshetty3416@gmail.com>

mergify[bot] committed:
fix: the YYYY.MM.DD format not working for the document naming rule (backport #15634) (#15645)

Co-authored-by: Rohit Waghchaure <rohitw1991@gmail.com>

mergify[bot] committed:
fix: document email in timeline (backport #15592) (#15646)

Co-authored-by: Raffael Meyer <14891507+barredterra@users.noreply.github.com>

mergify[bot] committed:
fix: customize same property for multiple rows of DocType Link / DocType Action (backport #15491) (#15647)

Co-authored-by: Pruthvi Patel <pruthvipatel145@gmail.com>

mergify[bot] committed:
fix: Recursive routing for Form route (backport #15539) (#15648)

Co-authored-by: Suraj Shetty <13928957+surajshetty3416@users.noreply.github.com>
Co-authored-by: Pruthvi Patel <pruthvipatel145@gmail.com>

lucasvazq committed:
fix: allow custom queues

ankush committed:
fix(patch): reload property setter for new fields (#15654)

surajshetty3416 committed:
fix: Set toast's background color based on alert type

(cherry picked from commit a7a265c4b1e04998cc949c0fa3cc6e49858366db)

mergify[bot] committed:
Merge pull request #15657 from gavindsouza/validate_link-dashboard

fix: validate_link calls via Edit Shortcut in Dashboard

mergify[bot] committed:
fix: validate_link calls via Edit Shortcut in Dashboard (backport #15657) (#15658)

Co-authored-by: Summayya <frappe@Summayyas-MacBook-Air.local>
Co-authored-by: Shariq Ansari <sharique.rik@gmail.com>

surajshetty3416 committed:
Merge pull request #15660 from frappe/mergify/bp/version-13-hotfix/pr-15659

rutwikhdev committed:
feat: patched frappe email to work with frappecloud mail app (#15652)

* feat: overrides default `smtp.sendmail` and sends the data to hook method if
  it's set

* feat: update email queue and communication status

* feat: override default sender details

gavindsouza committed:
fix: Fetch file extension from Response

(cherry picked from commit 04d95d9e044fc4e972ba4ef39d6f9c77b5c6227b)

gavindsouza committed:
chore: Add typing, use better supported APIs

(cherry picked from commit 36f58335a6851a8d0df354c6f5e5bf8036425231)

lucasvazq committed:
fix: remove site config and reformat the code

lucasvazq committed:
fix: remove redundant code

gavindsouza committed:
fix: if response content-type is bin, try guessing downloaded content
type

(cherry picked from commit f44e53b0e9c012313523270ab1f315ba28c9645a)

gavindsouza committed:
test: Add test for thumbnail creation of image without extn

# Conflicts:
#	frappe/core/doctype/file/test_file.py

(cherry picked from commit 57867ed7b356b8c7c4e885fcb0fa35cf0a0c0581)

mergify[bot] committed:
fix: Update colors of standard alert classes based on the new design (backport #15663) (#15664)

Co-authored-by: Suraj Shetty <surajshetty3416@gmail.com>
Co-authored-by: Suraj Shetty <13928957+surajshetty3416@users.noreply.github.com>

mergify[bot] committed:
Merge pull request #15653 from lucasvazq/version-13-hotfix

fix: allow custom queues

mergify[bot] committed:
fix: Set first day of the week for date range (backport #15669) (#15671)

Co-authored-by: Suraj Shetty <13928957+surajshetty3416@users.noreply.github.com>

mergify[bot] committed:
fix: Set first day of the week for date range (backport #15669) (#15672)

Co-authored-by: Suraj Shetty <13928957+surajshetty3416@users.noreply.github.com>

pruthvi145 committed:
test: add comments and test cases for `set_name_translation`

(cherry picked from commit a5096a1a816265a8140d11ac9bfe6e53247a5636)

hrwX committed:
chore: backport multistep_webforms

rohitwaghchaure committed:
chore: change log for v13.19.0

rohitwaghchaure committed:
Merge pull request #15689 from rohitwaghchaure/change-log-for-v13-19

chore: change log for v13.19.0

rohitwaghchaure committed:
bumped to version 13.19.0

rohitwaghchaure committed:
Merge branch 'version-13-pre-release' into version-13

Commit:
Merge remote-tracking branch 'upstream/version-13' into merge-v13

mergify[bot] committed:
feat: add translations to datatable (backport #15551) (#15567)

Co-authored-by: hrwx <himanshuwarekar@yahoo.com>
Co-authored-by: barredterra <14891507+barredterra@users.noreply.github.com>
Co-authored-by: Suraj Shetty <13928957+surajshetty3416@users.noreply.github.com>

mergify[bot] committed:
chore: add translation context to strings (backport #15649) (#15678)

Co-authored-by: barredterra <14891507+barredterra@users.noreply.github.com>

shariquerik committed:
fix: conflict in package.json

shariquerik committed:
Merge branch 'version-13-hotfix' into mergify/bp/version-13-hotfix/pr-15635

shariquerik committed:
Merge branch 'version-13-hotfix' into mergify/bp/version-13-hotfix/pr-15635

hrwX committed:
feat: time field in webform

(cherry picked from commit b550096161006b17890e430705ca2f779f08662d)

saxenabhishek committed:
fix: sbool converting int stored as string

(cherry picked from commit 0f70b0bccb1099e731d7611cbcda311fce6eb93a)

rmehta committed:
Merge pull request #15638 from frappe/mergify/bp/version-13-hotfix/pr-15635

build: Upgrade Frappe Gantt from 0.5.0 to 0.6.0 (backport #15635)

mergify[bot] committed:
fix: translate fieldname in report dialog and alert (backport #15682) (#15704)

Co-authored-by: barredterra <14891507+barredterra@users.noreply.github.com>

ruchamahabal committed:
fix(User Type): Select perms not updated for link fields in child table

(cherry picked from commit e88707a8d409bc248aee6f6cf88e347115a0a6fa)

gavindsouza committed:
feat(minor): Add "in list view" fields in sort selector

(cherry picked from commit 7fa9a3fd39cbc5ecc4db8da50317c3927e302271)

gavindsouza committed:
refactor(minor): SortSelector.setup_from_doctype

(cherry picked from commit 5742217e8e680e52edc456f59870443398222e74)

shariquerik committed:
test: fixed failing multi select dialog test

(cherry picked from commit 72352133c536e00958a4419322c62207953dda9e)

saxenabhishek committed:
fix: timedelta parsing in pypika

(cherry picked from commit d04091265b74a70f051db7662022627ca5d71c13)

surajshetty3416 committed:
ci: Run cypress tests on each re-run

- using $GITHUB_RUN_ID-$GITHUB_RUN_ATTEMPT to keep each re-run unique
- Previously, cypress used to just return "Run finished" (without actually re-running the test) after re-running the GA build (without updating the code). This used to give false impression that all tests are passed.

(cherry picked from commit da5779dc439c4dfc8445358ea7c84506128c9dfa)

# Conflicts:
#	.github/workflows/ui-tests.yml

surajshetty3416 committed:
ci: Resolve conflicts

surajshetty3416 committed:
Merge pull request #15719 from frappe/mergify/bp/version-13-hotfix/pr-15716

surajshetty3416 committed:
Merge pull request #15720 from frappe/mergify/bp/version-13-hotfix/pr-15710

surajshetty3416 committed:
Merge pull request #15712 from frappe/mergify/bp/version-13-hotfix/pr-15709

saxenabhishek committed:
test: avoid parameterization of qb objects

(cherry picked from commit 923651cd8fb7c6b3a473a3366c9561b3fd826a24)

hrwX committed:
feat: get_count, get_stats for virtual doctype

(cherry picked from commit 26a70e3cf1e3b5997ea30a6c329efc2af41e13e3)

gavindsouza committed:
fix(Custom DocPerm): Use Link type instead of Data for parent (#15715)

* fix: Use Link type instead of Data for parent

This makes sure frappe.rename_doc updates linked doctype records

* Revert "fix: Use Link type instead of Data for parent"

This reverts commit 93b37351069df9a79f04bfa326378f09bf97700f.

* fix(rename_doc): Handle re-linking Custom DocPerm records

(cherry picked from commit d78b49620863dec0975212f0d3e57fdd1698a96c)

surajshetty3416 committed:
Merge pull request #15725 from frappe/mergify/bp/version-13-hotfix/pr-15715

shariquerik committed:
test: flaky report_view fix

(cherry picked from commit 73f31737cd493541e571b65f6faca754a38cff98)

surajshetty3416 committed:
fix: Remove salutation from email

Reasons:
- It shows up only when few conditons (which are not very clear) are met. Since the behaviour is not consistent it creates unnecessary confusion
- Salutations are not configurable. Not all users want to start their email with "Dear" in which case they have to clear it first.

(cherry picked from commit a90a9ed807b9658a5e16cd5502fc5a7d02023d16)

hrwX committed:
chore: review changes

(cherry picked from commit dbb609bd0b86f52115992955d7e66407370c66c7)

surajshetty3416 committed:
Merge pull request #15736 from frappe/mergify/bp/version-13-hotfix/pr-15730

surajshetty3416 committed:
Merge pull request #15743 from frappe/mergify/bp/version-13-hotfix/pr-15740

shariquerik committed:
fix: Show image in sidebar in mobile view for item

(cherry picked from commit 0a243defd358be2fd81ea1bcd0f363a4dfc2f392)

tundebabzy committed:
fix(File): Handle file_path is None (#15552)

* fix: S3 automatic backup fails when the backup folder is empty

* Removed conditional that returns early when
`file_path` argument is invalid i.e `None` or empty string.

This means though that the function will throw an exception if it
receives such invalid input so devs should check their inputs before
calling the function

Co-authored-by: Mohammad Hasnain Mohsin Rajan <hasnain2808@gmail.com>

Co-authored-by: Mohammad Hasnain Mohsin Rajan <hasnain2808@gmail.com>
(cherry picked from commit 38caf3bb227fef01383aab2d1c6f1832c35eb392)

hasnain2808 committed:
chore: fix merge conflicts

hasnain2808 committed:
chore: remove attribute irrelevant for v13

mergify[bot] committed:
Merge pull request #15748 from frappe/mergify/bp/version-13-hotfix/pr-15568

fix(Customize Form): delete translation for renaming only if label is empty (backport #15568)

mergify[bot] committed:
Merge pull request #15749 from frappe/mergify/bp/version-13-hotfix/pr-15707

fix(User Type): Select perms not updated for link fields in child table (backport #15707)

mergify[bot] committed:
Merge pull request #15747 from frappe/mergify/bp/version-13-hotfix/pr-15552

fix(File): Handle file_path is None (backport #15552)

surajshetty3416 committed:
fix: Code block style

(cherry picked from commit 1569aa768c3ad001b72d663ca564d8e4d628f1a4)

# Conflicts:
#	frappe/public/scss/common/css_variables.scss

surajshetty3416 committed:
fix: Make text-editor resizable

(cherry picked from commit 0b2015f49890857ba9c17c85a3c28af10715053d)

surajshetty3416 committed:
fix: Resolve conflicts

surajshetty3416 committed:
Merge pull request #15760 from frappe/mergify/bp/version-13-hotfix/pr-15759

surajshetty3416 committed:
Merge pull request #15761 from frappe/mergify/bp/version-13-hotfix/pr-15752

surajshetty3416 committed:
Merge pull request #15731 from frappe/mergify/bp/version-13-hotfix/pr-15703

surajshetty3416 committed:
Merge pull request #15751 from frappe/mergify/bp/version-13-hotfix/pr-15601

gavindsouza committed:
fix(test): Add missing test image for test_make_thumbnail

(cherry picked from commit 734188313f28b67f8888e1b79e0f2c1bc7a570d9)

mergify[bot] committed:
chore: update translation for first day of week (#15766)

Added translation connected to: https://github.com/frappe/frappe/pull/15669

(cherry picked from commit 4718ebc2c31e77756c7988f29c3e04f6881bddbf)

Co-authored-by: Wolfram Schmidt <wolfram.schmidt@phamos.eu>

barredterra committed:
fix: translate title in msgprint

(cherry picked from commit 90b5664758f189ffcf73ff8217a4dd4f9f0e1541)

barredterra committed:
test: compare message instead of entire dict

(cherry picked from commit 20d9a7862812373e532cbdf833f73fc19dd83e56)

surajshetty3416 committed:
fix: Add patch to remove chat related doctypes

(cherry picked from commit 6f5feb05fabb87cccc05166495af9583aa304c69)

surajshetty3416 committed:
fix: Use capital M in max to avoid libsass compatibility issue

Similar to https://github.com/frappe/frappe/pull/14348

surajshetty3416 committed:
Merge pull request #15780 from frappe/fix-build

surajshetty3416 committed:
Merge branch 'version-13-hotfix' into mergify/bp/version-13-hotfix/pr-15777

surajshetty3416 committed:
Merge pull request #15778 from frappe/mergify/bp/version-13-hotfix/pr-15777

hrwX committed:
fix: show sidebar in nav for md and sm screen size

(cherry picked from commit ede975564992d040a350a263b17b37a08773a19a)

gavindsouza committed:
feat(minor): Case option in run-tests for specifying TestCase

(cherry picked from commit 6b160a0dbe6a55c2781833d128791d9bc3ca3337)

gavindsouza committed:
fix: Handle parsing and formatting timedeltas

* Added utils parse_timedelta, format_timedelta
* Added to json_handler for de-serializing timedelta objects

(cherry picked from commit 60f8bd1cfe3230b853864e0839cb534f615bb28b)

gavindsouza committed:
test: Added unit tests for format_timedelta, parse_timedelta,
json_handler

(cherry picked from commit 9205a4e06827136ab23b4a8db1335d69b4dd7f6e)

gavindsouza committed:
style: Sort imports

(cherry picked from commit 42365c3c906804fef69b5a1f544fa7b472b79135)

mergify[bot] committed:
Merge pull request #15788 from frappe/mergify/bp/version-13-hotfix/pr-15783

fix: show sidebar in nav for md and sm screen size (backport #15783)

barredterra committed:
Merge remote-tracking branch 'upstream/version-13-hotfix' into backport-multistep_webform

mergify[bot] committed:
Merge branch 'version-13-hotfix' into mergify/bp/version-13-hotfix/pr-15722

mergify[bot] committed:
Merge pull request #15683 from hrwX/backport-multistep_webform

chore: backport multistep webform (#14978)

surajshetty3416 committed:
Merge pull request #15770 from frappe/mergify/bp/version-13-hotfix/pr-15722

surajshetty3416 committed:
Merge pull request #15797 from frappe/mergify/bp/version-13-hotfix/pr-15768

gavindsouza committed:
fix: Avoid TypeError when guess_extension returns None

(cherry picked from commit 34d982ae68a2ea39ee8303482de1658f4c56a991)

gavindsouza committed:
fix: Format timedelta object accurately

(cherry picked from commit ca0341ebe8113096ddf7b29210290774dbca0ec6)

surajshetty3416 committed:
fix: Call validate_link only if "value" is passed

(cherry picked from commit 851e66e5643de3f8431c9e7b2d84064b3f8f5024)

mergify[bot] committed:
chore: Update de.csv (#15802) (#15806)

added missing translation in DocType User

(cherry picked from commit fadb07393ba75d1741dfd17f897b82a0002b9f19)

Co-authored-by: Wolfram Schmidt <wolfram.schmidt@phamos.eu>

surajshetty3416 committed:
test: Update UI test for link field

(cherry picked from commit 34e258c065bd4c83c6c2d89db02612f0b1b4eb89)

rohitwaghchaure committed:
chore: Merge branch 'version-13-hotfix' into 'version-13-pre-release'

rohitwaghchaure committed:
Merge pull request #15814 from rohitwaghchaure/merge-branch-hotfix-to-pre-release-for-13-20

chore: Merge branch 'version-13-hotfix' into 'version-13-pre-release'

surajshetty3416 committed:
test: Update test case for default value in link field

(cherry picked from commit 52604d9206402d2de5dfb46a7aa6e6437e8eb41f)

surajshetty3416 committed:
fix: Update set_nulls code to filters out empty values for link fields

- Empty values for link field is invalid and causes issue mentioned here https://github.com/frappe/frappe/pull/15320

(cherry picked from commit d2df8fa7f283f0bdb7888f7663f253b7585edd9f)

surajshetty3416 committed:
chore: Remove duplicate command

(cherry picked from commit c55b012e3d7a5aba3f55d3ebe9fc1f16ea92eb37)

surajshetty3416 committed:
fix: Update filter whilte removing undefined

(cherry picked from commit 6a1642e7c309de1b2c26bda2cefa0995ddf10074)

gavindsouza committed:
test: Add tests for get_time, get_timedelta

(cherry picked from commit 2114fc4ffbd6fbfcfe4a2a86c8d7a2c68f9804e6)

gavindsouza committed:
fix: Return correct types for time utils

(cherry picked from commit 33b1bb17511644d62a4a5180e3f8d604945376a5)

gavindsouza committed:
fix(test): Update test according to API behaviour

(cherry picked from commit 0ad369a812d1258d2ed920d7d7f3b53a42bba20c)

saxenabhishek committed:
fix: missing `param_wrapper` in recursive calls

(cherry picked from commit 0c5247f8502932cfd06f88274485e61797a7c5b1)

saxenabhishek committed:
test: string values inside qb objects

case inside update queries needs a recursive call param wrapper got lost

(cherry picked from commit a87139f9511749d647db1fab6e6b9ebb513a5e3e)

nabinhait committed:
fix: delete chat message attachments (#15852)

(cherry picked from commit 21b997bd8b1fec893e03ad0645be23fee8ce8b51)

surajshetty3416 committed:
Merge pull request #15860 from frappe/mergify/bp/version-13-pre-release/pr-15852

surajshetty3416 committed:
Merge pull request #15861 from frappe/mergify/bp/version-13-pre-release/pr-15856

hasnain2808 committed:
revert: "refactor: set amended docname to original docname" (#15872)

amend docs would use new names and cancelled docs would retain the orignal names

(cherry picked from commit 3d001168425b5a0e506c212e34aa0f4d62e75d39)

mergify[bot] committed:
Merge pull request #15881 from frappe/mergify/bp/version-13-pre-release/pr-15872

revert: "refactor: set amended docname to original docname" (backport #15872)

gavindsouza committed:
fix: Sync query_builder module with develop

The module looked suspiciouslly different from its state on
develop...more duplication - possibly due to an earlier botched cherry
pick

Also, ported https://github.com/frappe/frappe/pull/14457 from develop

(cherry picked from commit 243b02227920851be292de671704885bf00daf4e)

gavindsouza committed:
fix: Remove import re-definitions

(cherry picked from commit 4842ae47e55d3df3ebd14092ad5b9cae9aad4d42)

gavindsouza committed:
fix: Handle missing pluck functionality in frappe.db.get_values

(cherry picked from commit ad8b3adf601692aebab2d498def570c87a11f893)

gavindsouza committed:
fix: Revert ported aggregate qb functions

Had to undo since Query engine is develop only as of rn

Added via https://github.com/frappe/frappe/pull/15883/commits/243b02227920851be292de671704885bf00daf4e

(cherry picked from commit e155b2e49fe4c792572141942c42e8dfd119dae4)

gavindsouza committed:
fix: Check if binary file before not handling static file

Similar fix for develop's router via https://github.com/frappe/frappe/pull/15661/commits/9dbaf252f0acaac73ecb5181d0eba453dbcef973

(cherry picked from commit e8d699d40edbba475d10f1435933a70cde2ba910)

gavindsouza committed:
test(fix): Ignore case while checking for update query

ref: https://docs.python.org/3/library/stdtypes.html#str.casefold
(cherry picked from commit 269ea78d968c8362a59e854ebfb0a9a8a6a757ff)

gavindsouza committed:
Merge pull request #15896 from frappe/mergify/bp/version-13-pre-release/pr-15838

fix: Thumbnail for external images (from URL) [v13] (backport #15838)

gavindsouza committed:
test(fix): Use indexing instead of args key for mocked call

(cherry picked from commit 7b44440a9eb202ca42f472a03288b180b10259f5)

ankush committed:
fix: executing non-select qb code from whitelisted methods

Co-Authored-By: Gavin D'souza <gavin18d@gmail.com>
(cherry picked from commit c0090960929f23fa7a269877e3f10120ba2f99cb)

ankush committed:
test: restricted QB

(cherry picked from commit 2bbd83da3499920f5cafef4f67528f927a60936c)

gavindsouza committed:
feat: additional filter support (#14102)

Partial port of https://github.com/frappe/frappe/pull/14102

Co-authored-by: Aradhya <aradhyatripathi51@gmail.com>
Co-authored-by: Ankush <ankush@frappe.io>
(cherry picked from commit 84cb036642838dce581a9a32ae9c483076366473)

gavindsouza committed:
fix: Merge conflict

gavindsouza committed:
Merge pull request #15911 from frappe/mergify/bp/version-13-pre-release/pr-15883

refactor: Query Building & Data Handling [v13] (backport #15883)

rohitwaghchaure committed:
chore: change log for v13.20.0

rohitwaghchaure committed:
Merge pull request #15924 from rohitwaghchaure/change-log-for-v13-20

chore: change log for v13.20.0

rohitwaghchaure committed:
bumped to version 13.20.0

rohitwaghchaure committed:
Merge branch 'version-13-pre-release' into version-13

Commit:
Merge remote-tracking branch 'upstream/version-13' into merge-v13

gavindsouza committed:
fix: Check if valid file instead of path

(cherry picked from commit 10be9ac41be94fb6ea28a1118b004c08591a4b99)

gavindsouza committed:
Merge pull request #15937 from frappe/mergify/bp/version-13-pre-release/pr-15936

fix(render): Check if valid file instead of path [v13] (backport #15936)

rohitwaghchaure committed:
Merge branch 'version-13-pre-release' into version-13

rohitwaghchaure committed:
bumped to version 13.20.1

Commit:
Merge remote-tracking branch 'upstream/version-13' into merge-v13
